modules/javafx.graphics/src/main/java/com/sun/javafx/font/directwrite/DWFactory.java

```diff
@@ -116,11 +118,29 @@ private static void checkThread() {
         }
     }
 
+    private static void addGraphicsPipelineDisposeHook(Runnable runnable) {
+        try {
+            Class plc = Class.forName("com.sun.prism.GraphicsPipeline");
+            Method gpm = plc.getMethod("getPipeline", (Class[])null);
+            Object plo = gpm.invoke(null);
+            Method adhm = plc.getMethod("addDisposeHook", Runnable.class);
+            adhm.invoke(plo, runnable);
+        } catch (Exception e) {
+        }
+    }
+
     static synchronized IWICImagingFactory getWICFactory() {
         checkThread();
         /* Using single threaded WIC Factory as it should only be used by the rendering thread */
         if (WIC_FACTORY == null) {
             WIC_FACTORY = OS.WICCreateImagingFactory();
+            addGraphicsPipelineDisposeHook(() -> {
+                checkThread();
+                WIC_FACTORY.Release();
+                /* Uninitialize COM as it was initialized by WICCreateImagingFactory() */
+                OS.CoUninitialize();
```

Comment:
Do you think it would be cleaner to move this into `IWICImagingFactory`, by overriding the `Release` method, calling `super.Release()`, and then calling `OS.CoUninitialize()` from there? I don't have a strong opinion, so this is just a suggestion.

Reply:
@kevinrushforth 
> Do you think it would be cleaner to move this into `IWICImagingFactory`, by overriding the `Release` method, calling `super.Release()`, and then calling `OS.CoUninitialize()` from there?

From my perspective `IWICImagingFactory` just seems to be a light wrapper around the raw COM `IWICImagingFactory` pointer. Since with COM's C++ API calling `IWICImagingFactory::Release()` wouldn't call `CoUninitialize()`, I think we should mirror that here as well.

Actually, perhaps the `CoInitializeEx()` call should be moved out of `WICCreateImagingFactory` and made into its own `OS#CoIninitializeEx()` method? That way from the Java side it would be obvious that the `OS.CoUninitialize()` call is there to match the `OS.CoInitializeEx()` call.